Refactor crc32_fold functions into single crc32_copy

nmoinvaz opened the pull request:
X86 code was the only implementation that was using the fold methodology. Mainly it was to prevent having to do Barrett final reduction more than once. However, this added additional code complexity which may not be worth it.

The original implementation also had problems where the initial crc was not being XORed or XORed improperly.

I have also added missing crc32_copy functions for each platform to prevent an additional indirection call.

phprus commented on arch/generic/crc32_copy_c.c:
```diff
@@ -0,0 +1,14 @@
+#include "crc32.h"
+
+Z_INTERNAL uint32_t crc32_copy_c(uint32_t crc, uint8_t *dst, const uint8_t *src, size_t len) {
+    crc = FUNCTABLE_CALL(crc32)(crc, src, len);
```
This implementation makes a double indirect call via `FUNCTABLE_CALL` on arm and loongarch64.

nmoinvaz replied:
lol it's not used

phprus replied:
You are right!

phprus commented on functable.c:
```diff
@@ -358,10 +340,7 @@ static int init_functable(void) {
-    FUNCTABLE_VERIFY_ASSIGN(ft, crc32_fold_copy);
-    FUNCTABLE_VERIFY_ASSIGN(ft, crc32_fold_final);
-    FUNCTABLE_VERIFY_ASSIGN(ft, crc32_fold_reset);
+    FUNCTABLE_VERIFY_ASSIGN(ft, crc32_copy);
```
`ft.crc32_copy` may be NULL.
Using `FUNCTABLE_VERIFY_ASSIGN` will produce the error:
```
Zlib-ng functable failed initialization!
```

https://github.com/zlib-ng/zlib-ng/actions/runs/20088522280/job/57631057600?pr=2048#step:18:43

nmoinvaz replied:
I forgot to push my changes last night.

phprus commented on functable.c:
```diff
@@ -403,24 +381,16 @@ static uint32_t crc32_stub(uint32_t crc, const uint8_t* buf, size_t len) {
-    FUNCTABLE_INIT_ABORT;
-    return functable.crc32_fold_reset(crc);
+static uint32_t crc32_copy_stub(uint32_t crc, uint8_t *dst, const uint8_t *src, size_t len) {
+    if (functable.crc32_copy == crc32_copy_stub)
+        return crc32_copy_fallback(crc, dst, src, len);
+    return functable.crc32_copy(crc, dst, src, len);
```
`init_functable()` doesn't modify `functable.crc32_copy` and will always execute `crc32_copy_fallback` with double indirection on all architectures.
If we add an assignment to `functable.crc32_copy`, double indirection will occur on all platforms except x86.
Furthermore, the lack of a general version of the functions breaks the native build, since the `native_crc32_copy` macro is only present on x86 ( https://github.com/zlib-ng/zlib-ng/actions/runs/20107874162/job/57696841816?pr=2048#step:17:18 ).

nmoinvaz replied:
Yeah looks like there is no way to avoid the dummy functions all over the place.

nmoinvaz committed:
Merged crc32_fold functions save, load, reset

nmoinvaz committed:
Rename crc32_fold_pclmulqdq_tpl.h to crc32_pclmulqdq_tpl.h

nmoinvaz commented:
There was some bugs wen using init crc in crc32_pclmulqdq... but there appears to be more.. :-(

nmoinvaz committed:
Fix initial crc value loading in crc32_(v)pclmulqdq

In main function, alignment diff processing was getting in the way of XORing
the initial CRC, because it does not guarantee at least 16 bytes have been
loaded.

In fold_16, src data modified by initial crc XORing before being stored to dst.

nmoinvaz committed:
Move crc32_fold_s struct into x86 implementation.

nmoinvaz commented:
Microsoft Surface Book 4 with AVX512 (although this tests only pclmulqdq, the base code is the same). Benchmarks are hard to get consistent on Windows, but I post them just to show no major performance loss with crc hashing.
```
.\benchmark_zlib_old.exe --benchmark_filter="crc32/pclmulqdq*"
2025-12-15T17:29:35-08:00
Running .\benchmark_zlib_old.exe
Run on (8 X 2995 MHz CPU s)
CPU Caches:
  L1 Data 48 KiB (x4)
  L1 Instruction 32 KiB (x4)
  L2 Unified 1280 KiB (x4)
  L3 Unified 12288 KiB (x1)
------------------------------------------------------------------
Benchmark                        Time             CPU   Iterations
------------------------------------------------------------------
crc32/pclmulqdq/1             3.71 ns         3.58 ns    165925926
crc32/pclmulqdq/8             16.5 ns         16.4 ns     44800000
crc32/pclmulqdq/12            28.7 ns         27.7 ns     29866667
crc32/pclmulqdq/16            25.8 ns         23.1 ns     26352941
crc32/pclmulqdq/32            28.8 ns         26.4 ns     24888889
crc32/pclmulqdq/64            22.7 ns         22.9 ns     28000000
crc32/pclmulqdq/512           70.2 ns         69.8 ns     11200000
crc32/pclmulqdq/4096           358 ns          357 ns      1792000
crc32/pclmulqdq/32768         2905 ns         2905 ns       263529
crc32/pclmulqdq/262144       22254 ns        22496 ns        29867
crc32/pclmulqdq/4194304     372159 ns       359869 ns         1867

.\benchmark_zlib_new.exe --benchmark_filter="crc32/pclmulqdq*"
2025-12-15T17:29:47-08:00
Running .\benchmark_zlib_new.exe
Run on (8 X 2995 MHz CPU s)
CPU Caches:
  L1 Data 48 KiB (x4)
  L1 Instruction 32 KiB (x4)
  L2 Unified 1280 KiB (x4)
  L3 Unified 12288 KiB (x1)
------------------------------------------------------------------
Benchmark                        Time             CPU   Iterations
------------------------------------------------------------------
crc32/pclmulqdq/1             5.96 ns         5.94 ns    100000000
crc32/pclmulqdq/8             18.7 ns         18.4 ns     37333333
crc32/pclmulqdq/12            26.1 ns         26.1 ns     26352941
crc32/pclmulqdq/16            23.0 ns         23.0 ns     29866667
crc32/pclmulqdq/32            26.1 ns         26.1 ns     26352941
crc32/pclmulqdq/64            29.8 ns         28.9 ns     24888889
crc32/pclmulqdq/512           69.5 ns         69.8 ns      8960000
crc32/pclmulqdq/4096           373 ns          377 ns      1947826
crc32/pclmulqdq/32768         2816 ns         2787 ns       263529
crc32/pclmulqdq/262144       22664 ns        22461 ns        32000
crc32/pclmulqdq/4194304     366029 ns       368968 ns         1948
```

Dead2 commented on test/benchmarks/benchmark_crc32_copy.cc:
```diff
@@ -0,0 +1,120 @@
+         * while the subsequent one happened to not be. This is not to be advantageous to AVX512
+         * (indeed, all lesser SIMD implementations benefit from this aligned allocation), but to
+         * control the _consistency_ of the results */
+        random_ints_src = (uint32_t *)zng_alloc(MAX_RANDOM_INTS_SIZE);
```
**note:** zng_alloc() no longer does alignment, might as well use malloc directly().
zng_alloc_aligned() does do aligned allocs however.

But since we add the misalignment treatment below, I don't think we actually need an aligned buffer to begin with.
Btw, since this file is not marked as renamed in git, it's hard to compare changes.

nmoinvaz replied:
For whatever reason, even when I try to do a rename in one commit and then edit in another, the GitHub website doesn't recognize it as two separate things.

nmoinvaz replied:
The change to zng_alloc was something suggested by Coderabbit, which noticed that only in this benchmark, was it not using the same as the others. This benchmark was using malloc, whereas the others are using zng_alloc.

Dead2 replied:
Yeah, I'd like to stop using zng_alloc in the tests, as it serves no real purpose currently and having the tests depend on it is making some things more complicated than need be (for example using the same test for stock zlib).

I was originally thinking we could have an alloc function that is able to both provide aligned and unaligned buffers, but I think using the misalignment offsets in the test is really a better solution.
The misalignment makes sure we don't just test/bench one specific case that just so happens to usually be the best-case, especially when that is not reflected in the real use of the function being tested/benchmarked having to work on random memory offsets.

You could say that I consider it a TODO to rid the tests/benchmarks of zng_alloc usage, and also introduce misalignment more places. This test was kind of my prototype for that, and I haven't seen or heard anything negative about it.

mtl1979 replied:
Consistency is always preferred as it will avoid confusing AI and new contributors... As the wrapper functions are inlined in Release builds, I see no reason currently to replace all calls to zng_alloc/zng_free with malloc/free in our own tests.

mtl1979 replied:
> For whatever reason, even when I try to do a rename in one commit and then edit in another, the GitHub website doesn't recognize it as two separate things.

If too much is edited, it might just ignore the fact that the file is renamed, as git internally doesn't support renames, it's just a metadata change. There might be a threshold for deciding when two files are separate, but knowing that doesn't make us any wiser.

Dead2 replied:
I'd say it is more confusing for both AI and contributors that we use a special alloc function that does nothing other than call malloc. Calling malloc directly is the cleanest and easiest way to do it, and should be easily understood by all C developers without having to look it up.

nmoinvaz replied:
I will remove the last commit and try to break up the previous changes to the tests into two separate commits but I don't think it will make a difference with regard to why GitHub shows.

nmoinvaz replied:
I have remvoed the last commit.

Dead2 commented on test/benchmarks/benchmark_crc32_copy.cc:
```diff
@@ -0,0 +1,120 @@
+}
+
+#define MAX_RANDOM_INTS (1024 * 1024)
+#define MAX_RANDOM_INTS_SIZE (MAX_RANDOM_INTS * sizeof(uint32_t))
```
**question:** Not sure why you replaced this?
IMO the previous BUFSIZE was easier to read and keep correct.
You have also increased the buffer from 32KB+32B to 4MB, but the tests only ever use up to 32KB+head and tail padding.

mtl1979 replied:
I thought using larger buffer allows benchmarking different Chorba implementations... I guess the bigger sizes are missing below...

Dead2 replied:
For the crc32-copy implementations, there is no need to test above 32k, because we only use those functions for up to window-sized data, unless I am missing something.

mtl1979 replied:
Only reason I could think of was to compare speeds of non-copying and copying variants as copying can cause extra overhead in addition to actual memcpy call(s)... 32-bit PowerPC (and 32-bit ARM before ARMv6) is pretty much only architecture I know that can't do fast copying inline with crc32 (or adler32) calculation as it doesn't support unaligned loads and stores.

nmoinvaz replied:
I have removed this commit.

Dead2 commented on test/test_crc32_copy.cc:
```diff
@@ -0,0 +1,88 @@
+/* test_crc32_copy.cc -- test for crc32 implementations while copying
```
**note:** Same thing here, old file is deleted and new created, instead of git knowing it was renamed. Not sure what was changed.

nmoinvaz replied:
I have split the rename up, hope that helps, if not then it is beyond my control.

Dead2 commented on test/test_crc32_copy.cc:
```diff
@@ -0,0 +1,88 @@
+    TEST_CRC32_COPY(chorba_sse41, crc32_copy_chorba_sse41, test_cpu_features.x86.has_sse41)
+#    endif
+#  endif
+    // Optimized functions
```
**suggestion:** This comment should probably go up at line 59, before the WITHOUT_CHORBA. I consider the Chorba ones to be optimizations compared to the braid one.

nmoinvaz replied:
Fixed.

nmoinvaz committed:
Refactor crc32_fold functions into single crc32_copy

nmoinvaz commented:
I do have plans in a future PR to speed up the final Barrett reduction using method described in Mattermost.

nmoinvaz committed:
Update crc32_fold test and benchmarks for crc32_copy

nmoinvaz commented on test/test_crc32_copy.cc:
```diff
@@ -0,0 +1,86 @@
+#  endif
+#  ifdef X86_PCLMULQDQ_CRC
+    // Is 16 bytes len the minimum for pclmul functions?
+    TEST_CRC32_COPY(pclmulqdq, crc32_copy_pclmulqdq, test_cpu_features.x86.has_pclmulqdq)
```
I will remove the old comment.

Dead2 commented:
I had another look, and I have no requests for changes at this point.
I do want to do some benchmarking though, but I'll have to wait for my migraine to subside, hopefully before the Christmas celebrations start.

Dead2 commented:
This needs a rebase

phprus commented on test/benchmarks/benchmark_crc32_copy.cc:
```diff
@@ -0,0 +1,105 @@
+    uint8_t *dstbuf;
+
+public:
+    void SetUp(const ::benchmark::State& state) {
```
Replace ``const ::benchmark::State& state`` with ``const ::benchmark::State&`` to fix warning: unused parameter 'state'.

phprus commented on test/benchmarks/benchmark_crc32_copy.cc:
```diff
@@ -0,0 +1,105 @@
+        benchmark::DoNotOptimize(crc);
+    }
+
+    void TearDown(const ::benchmark::State& state) {
```
Replace ``const ::benchmark::State& state`` with ``const ::benchmark::State&`` to fix warning: unused parameter 'state'.

Dead2 commented:
AMD 8700GE

# Develop
```
   text    data     bss     dec     hex filename
 155420    1352       8  156780   2646c libz-ng.so.2
 
----------------------------------------------------------------------
Benchmark                            Time             CPU   Iterations
----------------------------------------------------------------------
crc32/pclmulqdq/1                 2.00 ns         1.99 ns   1059542948
crc32/pclmulqdq/8                 13.0 ns         13.0 ns    161984104
crc32/pclmulqdq/12                19.3 ns         19.3 ns    109055824
crc32/pclmulqdq/16                11.2 ns         11.2 ns    188273065
crc32/pclmulqdq/32                12.5 ns         12.5 ns    168059757
crc32/pclmulqdq/64                14.8 ns         14.8 ns    141839433
crc32/pclmulqdq/512               31.2 ns         31.2 ns     67263154
crc32/pclmulqdq/4096               174 ns          174 ns     12117600
crc32/pclmulqdq/32768             1326 ns         1325 ns      1585935
crc32/pclmulqdq/262144           10541 ns        10527 ns       199474
crc32/pclmulqdq/4194304         170200 ns       169869 ns        12375
crc32/vpclmulqdq/1                1.98 ns         1.98 ns   1055497273
crc32/vpclmulqdq/8                13.0 ns         13.0 ns    161832047
crc32/vpclmulqdq/12               19.3 ns         19.3 ns    108861194
crc32/vpclmulqdq/16               10.3 ns         10.3 ns    203476214
crc32/vpclmulqdq/32               11.4 ns         11.4 ns    184022842
crc32/vpclmulqdq/64               13.1 ns         13.1 ns    160794055
crc32/vpclmulqdq/512              21.1 ns         21.1 ns     99724556
crc32/vpclmulqdq/4096             64.7 ns         64.6 ns     32572517
crc32/vpclmulqdq/32768             419 ns          419 ns      5016691
crc32/vpclmulqdq/262144           3254 ns         3250 ns       646896
crc32/vpclmulqdq/4194304         52948 ns        52841 ns        39703
crc32_fc/braid_c/16               25.6 ns         25.6 ns     82121310
crc32_fc/braid_c/48               69.5 ns         69.4 ns     30271939
crc32_fc/braid_c/192               134 ns          134 ns     15617683
crc32_fc/braid_c/512               220 ns          220 ns      9557227
crc32_fc/braid_c/4096             1165 ns         1163 ns      1806570
crc32_fc/braid_c/16384            4568 ns         4562 ns       460131
crc32_fc/braid_c/32768            9074 ns         9062 ns       231802
crc32_fc/chorba_sse2/16           25.6 ns         25.6 ns     82107931
crc32_fc/chorba_sse2/48           69.5 ns         69.4 ns     30264139
crc32_fc/chorba_sse2/192           123 ns          122 ns     17118692
crc32_fc/chorba_sse2/512           155 ns          155 ns     13603526
crc32_fc/chorba_sse2/4096          522 ns          522 ns      4035513
crc32_fc/chorba_sse2/16384        1818 ns         1816 ns      1156661
crc32_fc/chorba_sse2/32768        3643 ns         3638 ns       577364
crc32_fc/chorba_sse41/16          25.7 ns         25.7 ns     82020721
crc32_fc/chorba_sse41/48          69.4 ns         69.3 ns     30301732
crc32_fc/chorba_sse41/192          122 ns          122 ns     17169255
crc32_fc/chorba_sse41/512          154 ns          154 ns     13618203
crc32_fc/chorba_sse41/4096         522 ns          521 ns      4027124
crc32_fc/chorba_sse41/16384       1306 ns         1304 ns      1614963
crc32_fc/chorba_sse41/32768       2341 ns         2338 ns       896861
crc32_fc/pclmulqdq/16             8.93 ns         8.91 ns    233431288
crc32_fc/pclmulqdq/48             9.75 ns         9.74 ns    213635608
crc32_fc/pclmulqdq/192            15.0 ns         15.0 ns    139856890
crc32_fc/pclmulqdq/512            30.6 ns         30.6 ns     68653182
crc32_fc/pclmulqdq/4096            208 ns          207 ns     10124826
crc32_fc/pclmulqdq/16384           821 ns          820 ns      2558002
crc32_fc/pclmulqdq/32768          1637 ns         1635 ns      1283361
crc32_fc/vpclmulqdq/16            9.16 ns         9.15 ns    229229442
crc32_fc/vpclmulqdq/48            10.2 ns         10.1 ns    207154319
crc32_fc/vpclmulqdq/192           16.1 ns         16.0 ns    130979063
crc32_fc/vpclmulqdq/512           28.1 ns         28.0 ns     74652255
crc32_fc/vpclmulqdq/4096          99.0 ns         98.8 ns     21196222
crc32_fc/vpclmulqdq/16384          281 ns          281 ns      7472748
crc32_fc/vpclmulqdq/32768          492 ns          491 ns      4275693

```

# PR
```
   text    data     bss     dec     hex filename
 149100    1328       8  150436   24ba4 libz-ng.so.2

------------------------------------------------------------------------
Benchmark                              Time             CPU   Iterations
------------------------------------------------------------------------
crc32/pclmulqdq/1                   1.98 ns         1.98 ns   1057333762
crc32/pclmulqdq/8                   13.0 ns         13.0 ns    162024369
crc32/pclmulqdq/12                  19.2 ns         19.2 ns    109186431
crc32/pclmulqdq/16                  9.88 ns         9.88 ns    212742812
crc32/pclmulqdq/32                  10.8 ns         10.8 ns    195669008
crc32/pclmulqdq/64                  13.0 ns         13.0 ns    163306056
crc32/pclmulqdq/512                 30.1 ns         30.1 ns     69657284
crc32/pclmulqdq/4096                 174 ns          174 ns     12043432
crc32/pclmulqdq/32768               1339 ns         1339 ns      1568413
crc32/pclmulqdq/262144             10672 ns        10672 ns       196725
crc32/pclmulqdq/4194304           171686 ns       171686 ns        12248
crc32/vpclmulqdq/1                  1.97 ns         1.97 ns   1062151894
crc32/vpclmulqdq/8                  13.0 ns         13.0 ns    161799805
crc32/vpclmulqdq/12                 19.2 ns         19.2 ns    109101994
crc32/vpclmulqdq/16                 6.76 ns         6.76 ns    308683337
crc32/vpclmulqdq/32                 8.09 ns         8.09 ns    259520959
crc32/vpclmulqdq/64                 9.79 ns         9.79 ns    217432696
crc32/vpclmulqdq/512                18.5 ns         18.5 ns    110483587
crc32/vpclmulqdq/4096               60.6 ns         60.6 ns     34477891
crc32/vpclmulqdq/32768               413 ns          413 ns      5082507
crc32/vpclmulqdq/262144             3242 ns         3242 ns       648058
crc32/vpclmulqdq/4194304           52498 ns        52497 ns        40071
crc32_copy/braid/16                 25.5 ns         25.5 ns     82293481
crc32_copy/braid/48                 69.5 ns         69.5 ns     30242127
crc32_copy/braid/192                 135 ns          135 ns     15570234
crc32_copy/braid/512                 219 ns          219 ns      9603214
crc32_copy/braid/4096               1161 ns         1161 ns      1808660
crc32_copy/braid/16384              4558 ns         4558 ns       461073
crc32_copy/braid/32768              9047 ns         9047 ns       232594
crc32_copy/chorba_sse2/16           25.5 ns         25.5 ns     82294303
crc32_copy/chorba_sse2/48           69.5 ns         69.5 ns     30183946
crc32_copy/chorba_sse2/192           122 ns          122 ns     17159973
crc32_copy/chorba_sse2/512           153 ns          153 ns     13711437
crc32_copy/chorba_sse2/4096          522 ns          522 ns      4033218
crc32_copy/chorba_sse2/16384        1815 ns         1815 ns      1157240
crc32_copy/chorba_sse2/32768        3633 ns         3633 ns       577786
crc32_copy/chorba_sse41/16          25.6 ns         25.6 ns     82123497
crc32_copy/chorba_sse41/48          69.3 ns         69.3 ns     30321804
crc32_copy/chorba_sse41/192          122 ns          122 ns     17140094
crc32_copy/chorba_sse41/512          153 ns          153 ns     13744239
crc32_copy/chorba_sse41/4096         521 ns          521 ns      4028723
crc32_copy/chorba_sse41/16384       1296 ns         1296 ns      1619306
crc32_copy/chorba_sse41/32768       2336 ns         2336 ns       892637
crc32_copy/pclmulqdq/16             24.5 ns         24.5 ns     85779737
crc32_copy/pclmulqdq/48             28.4 ns         28.4 ns     73867265
crc32_copy/pclmulqdq/192            31.7 ns         31.7 ns     65784343
crc32_copy/pclmulqdq/512            43.5 ns         43.5 ns     48341254
crc32_copy/pclmulqdq/4096            220 ns          220 ns      9549300
crc32_copy/pclmulqdq/16384           850 ns          850 ns      2470313
crc32_copy/pclmulqdq/32768          1658 ns         1658 ns      1267600
crc32_copy/vpclmulqdq/16            24.4 ns         24.4 ns     86057241
crc32_copy/vpclmulqdq/48            25.6 ns         25.6 ns     82190198
crc32_copy/vpclmulqdq/192           28.7 ns         28.7 ns     72757292
crc32_copy/vpclmulqdq/512           39.6 ns         39.6 ns     52929780
crc32_copy/vpclmulqdq/4096          77.0 ns         77.0 ns     27255598
crc32_copy/vpclmulqdq/16384          360 ns          360 ns      5848394
crc32_copy/vpclmulqdq/32768          574 ns          574 ns      3658963
```

~6KB smaller library size.
However, the [v]pclmul crc32_copy functions are significantly slower. The others seem fine however.

nmoinvaz commented:
Sorry, I forgot to mention this, but the benchmark test in `develop` is not equal to the one in `PR`. They are testing different things: in `develop` it doesn't test `reset` or `final` as part of the benchmark loop.
https://github.com/zlib-ng/zlib-ng/blob/7be4c57f00d004d7fc3721138eb4e2619dd0a66e/test/benchmarks/benchmark_crc32_fold_copy.cc#L77

If you include in the benchmark loop the results will become closer to one another.

When testing compression or decompression using `crc32_copy` in `develop` versus `PR`, I predict you might notice a very slight performance decrease in it - the additional overhead of a some align/finalize instructions for each `crc32_copy` call that wasn't done before. But also, now it no longer has to do the functable jump for `reset` or `final`.

nmoinvaz commented:
Rebased

KungFuJesus commented:
This shouldn't make the binary smaller, it should compile to two different functions. If it doesn't, it's compiling a branch in per loop iteration that we don't want. I'd make sure it's not doing that.

nmoinvaz commented:
The size difference could be from the removal of reset and final functions in functable and elsewhere. And from the removal of crc32_fold_s struct and its helper functions. And the small copying is now moved into the template function.

Dead2 commented:
Please have a look at https://github.com/zlib-ng/zlib-ng/commit/ed3766eea9bb9c1f2c64f0acff1cb4ec797cc81c
That is my suggestion to changes on top of this PR.

```
        text    data     bss     dec     hex filename
Devel 154996    1352       8  156356   262c4 libz-ng.so.2	 
PR    148612    1328       8  149948   249bc libz-ng.so.2
Dead2 151556    1328       8  152892   2553c libz-ng.so.2

Devel: crc32_fc/vpclmulqdq/32768            492 ns          491 ns      4275693
PR:    crc32_copy/vpclmulqdq/32768          574 ns          574 ns      3658963
Dead2: crc32_copy/vpclmulqdq/32768          464 ns          464 ns      4534103
```

### Full results for my suggestion:
```
crc32/pclmulqdq/1                   1.98 ns         1.97 ns   1064505846
crc32/pclmulqdq/8                   13.0 ns         13.0 ns    161999348
crc32/pclmulqdq/12                  19.3 ns         19.3 ns    109050978
crc32/pclmulqdq/16                  9.68 ns         9.67 ns    217150362
crc32/pclmulqdq/32                  10.8 ns         10.7 ns    194937509
crc32/pclmulqdq/64                  12.9 ns         12.9 ns    163170329
crc32/pclmulqdq/512                 30.2 ns         30.2 ns     69614025
crc32/pclmulqdq/4096                 172 ns          172 ns     12238686
crc32/pclmulqdq/32768               1313 ns         1311 ns      1601385
crc32/pclmulqdq/262144             10457 ns        10443 ns       201178
crc32/pclmulqdq/4194304           168581 ns       168250 ns        12481
crc32/vpclmulqdq/1                  1.98 ns         1.97 ns   1061021542
crc32/vpclmulqdq/8                  13.0 ns         13.0 ns    161535477
crc32/vpclmulqdq/12                 19.3 ns         19.3 ns    108993842
crc32/vpclmulqdq/16                 6.94 ns         6.93 ns    302865940
crc32/vpclmulqdq/32                 8.10 ns         8.09 ns    259487959
crc32/vpclmulqdq/64                 9.53 ns         9.52 ns    219292891
crc32/vpclmulqdq/512                18.9 ns         18.8 ns    111689156
crc32/vpclmulqdq/4096               59.4 ns         59.3 ns     35413600
crc32/vpclmulqdq/32768               413 ns          413 ns      5086187
crc32/vpclmulqdq/262144             3253 ns         3248 ns       646195
crc32/vpclmulqdq/4194304           52520 ns        52412 ns        40061

crc32_copy/pclmulqdq/16             24.4 ns         24.3 ns     86186884
crc32_copy/pclmulqdq/48             26.7 ns         26.6 ns     78726904
crc32_copy/pclmulqdq/192            30.6 ns         30.6 ns     68740892
crc32_copy/pclmulqdq/512            42.8 ns         42.8 ns     49014509
crc32_copy/pclmulqdq/4096            220 ns          219 ns      9580582
crc32_copy/pclmulqdq/16384           846 ns          844 ns      2486594
crc32_copy/pclmulqdq/32768          1660 ns         1657 ns      1265864
crc32_copy/vpclmulqdq/16            24.5 ns         24.5 ns     85842244
crc32_copy/vpclmulqdq/48            25.5 ns         25.4 ns     82882307
crc32_copy/vpclmulqdq/192           28.9 ns         28.9 ns     72676658
crc32_copy/vpclmulqdq/512           39.9 ns         39.9 ns     52897140
crc32_copy/vpclmulqdq/4096          79.0 ns         78.9 ns     26635434
crc32_copy/vpclmulqdq/16384          249 ns          249 ns      8442736
crc32_copy/vpclmulqdq/32768          464 ns          464 ns      4534103
```

As you can see, the short lengths still take a performance hit from the reset and final now being included in the benchmark. But the longer lengths are no longer slower, in fact for vpclmul they are even faster than before. pclmul is more or less unchanged.

It looks to me like the vpclmul functions were too big for complete inlining, so it didn't end up making two big functions and instead had a common one that contained all the COPY branches. The pclmul code is smaller, so I'd guess it managed to split that up like it should.

KungFuJesus commented:
I'd maybe try forcing the inlining with our macro and measure how it does

Dead2 commented:
> I'd maybe try forcing the inlining with our macro and measure how it does

I already tested that, doing that after this commit resulted in identical size and speed (I figure it might be better to do that, but didn't want to make more changes than I needed to).

Doing it just on top of the PR results in something like a half-measure, it is better but not as good for some reason.
Could it be that it becomes too much for the compiler to track, so it loses out on some optimization, but with COPY being a define instead of a variable passed to the function, it manages better? Just my speculation, I didn't look closely at it, I just wasn't that satisfied and went on to rewrite it.

nmoinvaz committed:
Handle initial crc only at the beginning of crc32_(v)pclmulqdq

nmoinvaz commented:
Interesting. So the whole `const int COPY` code pattern doesn't work if the function is too big?

In my second to last commit, I fixed vpclmulqdq where I wasn't checking if crc != 0 before doing initial crc XOR.

I have made some changes to initial crc handling to remove it from all of the loops. I am only able to easily do this because of prior commits in my PR.

nmoinvaz committed:
Move fold calls closer to last change in xmm_crc# variables.

nmoinvaz commented:
@dead2, I feel like that patch is a step backwards. I do have some other changes that might result in shorter function that could fix the problem, what do you think about that?

nmoinvaz committed:
Move COPY out of fold_16 inline with other fold_# functions.

nmoinvaz committed:
Use tail optimization in final barrett reduction

Fold 4x128-bit into a single 128-bit value using k1/k2 constants, then reduce
128-bits to 32-bits.

https://www.corsix.org/content/alternative-exposition-crc32_4k_pclmulqdq

nmoinvaz commented:
I put a few of the commits in here so you can see. There is another few commits probably after these too. The code is becoming more simplified..

nmoinvaz commented:
@Dead2 I wonder as an alternative to your patch, if I can just split the crc32_vpclmulqdq and crc32_copy_vpclmulqdq to be defined in two separate .c source files and crc32_pclmulqdq_tpl.h to be included in both files.

Dead2 commented:
> @Dead2 I wonder as an alternative to your patch, if I can just split the crc32_vpclmulqdq and crc32_copy_vpclmulqdq to be defined in two separate .c source files and crc32_pclmulqdq_tpl.h to be included in both files.

I'd be suspicious to think that might just be catering to compiler heuristics that might vary/change without notice.
I'll try to benchmark your suggestions when I can, I am all for it, my commit was meant mostly to highlight the problem.

I do think you should use z_forceinline on the crc32_impl function, and also add inline to all the static functions. Without those, the compiler is less likely to inline them on its own initiative.

nmoinvaz committed:
Add static inline/Z_FORCEINLINE to crc32_(v)pclmulqdq functions.

nmoinvaz commented:
Doing `static inline` and `Z_FORCEINLINE` increases the size ` 183808 minigzip.exe -> 187904 minigzip.exe` so it's likely better inlining for performance (at least for MSVC 17.13.2).

Dead2 commented:
8700GE
```
   text    data     bss     dec     hex filename
 151012    1328       8  152348   2531c libz-ng.so.2
```
```
crc32/pclmulqdq/1                   1.98 ns         1.98 ns   1062418251
crc32/pclmulqdq/8                   13.0 ns         13.0 ns    161965298
crc32/pclmulqdq/12                  19.3 ns         19.3 ns    108891207
crc32/pclmulqdq/16                  6.49 ns         6.48 ns    324039546
crc32/pclmulqdq/32                  7.40 ns         7.39 ns    284007036
crc32/pclmulqdq/64                  9.41 ns         9.40 ns    223302157
crc32/pclmulqdq/512                 29.5 ns         29.5 ns     71093253
crc32/pclmulqdq/4096                 170 ns          170 ns     12351123
crc32/pclmulqdq/32768               1311 ns         1310 ns      1603974
crc32/pclmulqdq/262144             10449 ns        10435 ns       201222
crc32/pclmulqdq/4194304           168845 ns       168511 ns        12464
crc32/vpclmulqdq/1                  1.99 ns         1.99 ns   1058549198
crc32/vpclmulqdq/8                  13.0 ns         13.0 ns    161630698
crc32/vpclmulqdq/12                 19.3 ns         19.3 ns    108961569
crc32/vpclmulqdq/16                 5.71 ns         5.71 ns    367779872
crc32/vpclmulqdq/32                 6.99 ns         6.98 ns    300216984
crc32/vpclmulqdq/64                 8.55 ns         8.54 ns    245781237
crc32/vpclmulqdq/512                23.6 ns         23.5 ns     89266928
crc32/vpclmulqdq/4096               64.2 ns         64.2 ns     32745801
crc32/vpclmulqdq/32768               419 ns          418 ns      5020346
crc32/vpclmulqdq/262144             3256 ns         3251 ns       644611
crc32/vpclmulqdq/4194304           52689 ns        52582 ns        39973
```
```
crc32_copy/pclmulqdq/16             24.3 ns         24.2 ns     86583633
crc32_copy/pclmulqdq/48             25.3 ns         25.3 ns     83285002
crc32_copy/pclmulqdq/192            28.9 ns         28.9 ns     72682497
crc32_copy/pclmulqdq/512            42.3 ns         42.2 ns     49681085
crc32_copy/pclmulqdq/4096            219 ns          218 ns      9603296
crc32_copy/pclmulqdq/16384           835 ns          834 ns      2519484
crc32_copy/pclmulqdq/32768          1654 ns         1652 ns      1271778
crc32_copy/vpclmulqdq/16            24.4 ns         24.3 ns     85948869
crc32_copy/vpclmulqdq/48            24.4 ns         24.4 ns     86404568
crc32_copy/vpclmulqdq/192           27.7 ns         27.7 ns     75855975
crc32_copy/vpclmulqdq/512           40.1 ns         40.1 ns     52537441
crc32_copy/vpclmulqdq/4096          79.0 ns         78.9 ns     26600261
crc32_copy/vpclmulqdq/16384          251 ns          250 ns      8388627
crc32_copy/vpclmulqdq/32768          474 ns          473 ns      4435711
```

This is even faster than before for non-copy crc32 in the 16-512bytes benchmarks.
The copy variant is slower for the same functions, by about the same amount.
The vpclmul copy variant is faster on the bigger sizes, performing much like my suggestion did.

Overall this looks good to me, even with the slightly slower small-copy.
